kong/plugins/key-auth/handler.lua

```diff
@@ -179,9 +237,44 @@ local function do_authentication(conf)
     return nil, { status = 500, message = "An unexpected error occurred" }
   end
 
-  set_consumer(consumer, credential)
+  if conf.validate_signature then
+    -- 4. Verify sha256
+    local secret = credential.secret
+    local now = math.floor(ngx.time())
+    local sha = sha256:new()
+    local ks = key .. secret
+
+
+    local verify_sig = function(time)
+      sha:reset()
+      sha:update(ks)
+      sha:update(tostring(time))
+      local final = to_hex(sha:final())
+      --print("  verifying - sha(" .. ks .. " .. " .. tostring(time) .. ") = " .. final)
+      -- print(signature .. " == " .. final)
+      return signature == final
+    end
 
-  return true
+    if verify_sig(now) then
+      -- authenticated by the current second
+      set_consumer(consumer, credential)
+      return true
+    end
+
+    for distance = 1, 300 do
+      if verify_sig(now + distance) or verify_sig(now - distance) then
+        print(" ****  Verified")
+        -- authenticated with `distance` seconds
+        set_consumer(consumer, credential)
+        return true
+      end
+    end
+
+    return nil, { status = 500, message = "Invalid signature" }
+  else
```

Comment:
style: please avoid `else` branches if the previous branch already returns, as enforced in [conditional expressions](https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#conditional-expressions) throughout the codebase.